### PR TITLE
fix: final model of a single drug model is trained on all drugs

### PR DESCRIPTION
**Pull request by jfrog64**
# PR Checklist for all PRs

- [x] This comment contains a description of changes (with reason)
- [ ] Referenced issue is linked — no existing issue, this turned up while looking for saved single drug models in a results directory
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [ ] Documentation in `docs` is updated — not needed, `drevalpy.experiment` is already an `automodule` and the new parameter is documented in the `train_final_model` docstring

### Bug fixes

With `final_model_on_full_data`, the final model of a **single drug model is fitted on the responses of all
drugs**, and that model is then saved under `<model>/drugs/<drug_id>/final_model/`.

Where it comes from: `make_model_list` creates one entry per drug for a single drug model, so the loop body —
including the `train_final_model` call — runs once per drug. The CV path masks the data for that drug in
`get_datasets_from_cv_split`:

```python
if model_name in SINGLE_DRUG_MODEL_FACTORY.keys():
    output_mask = train_dataset.drug_ids == drug_id
    train_dataset.mask(output_mask)
    ...
```

but `train_final_model` is handed `response_data.copy()` and never learns which drug it is training for. It
does not fail, it just trains on everything, so each per-drug directory ends up holding a drug-agnostic model.
With `hyperparameter_tuning=True` the same applies to the tuning that precedes it: it is repeated per drug on
the full dataset.

Observed on a two-drug subset of GDSC1 (1920 responses, LCO, 2 splits, no tuning) with
`models=[SingleDrugRandomForest]`:

```
Running SingleDrugRandomForest.9858940
...
Training final model with application-specific validation strategy ...
Reduced training dataset from 1920 to 1874, due to missing features   # <- both drugs
```

while the CV steps of the same run trained on ~480 responses per drug and split. Checking the two saved
forests confirms it: the root node of the first tree holds 339 and 347 samples, which is what
`max_samples=0.2` gives for 1874 training rows — one drug alone (~937 rows) would give ~170.

Fix: `train_final_model` takes the `drug_id` that the pipeline already has and reduces the dataset the same way
the CV path does. Passing no `drug_id` for a single drug model now raises instead of silently pooling the
drugs. Multi-drug models are unaffected, and the `model_class not in baselines` condition is untouched —
baselines not getting a final model is a separate, deliberate decision.

Tests: `tests/test_final_model_single_drug.py`

* the final model of each drug is trained on that drug's responses only, and is saved
* a single drug model without a `drug_id` raises `ValueError`
* a multi-drug model still sees all drugs
* the call site in `drug_response_experiment` hands over the drug id of the current model, once per drug

The tests use small recording models that build their features in memory, so they do not need a data
directory; only the wiring test loads TOYv1 because the pipeline always adds `NaiveMeanEffectsPredictor`.
All four fail without the fix.

One thing I left alone: `train_final_model` is annotated `response_transformation: TransformerMixin`, but
`drug_response_experiment` passes `TransformerMixin | None` (default `None`) straight through. That is
harmless at runtime but would trip typeguard for a run without a response transformation. Happy to widen the
annotation here if you'd rather have it in this PR.

### CODE REVIEW

**Review by JudithBernett — APPROVED** [review]
oh no, thanks for spotting this. This is also an issue in cli_model_testing.py/run_final_split. Working on it rn
